### ISSUE TITLE
Media: SQL DELETE fails but files are unlinked

### ISSUE THREAD

**Issue by R3dLantern** [sulu/sulu#6590]
| Q | A
| --- | ---
| Sulu Version | 2.4.2
| PHP Version | 8.0.13
| DB Version | MariaDB 10.4.12
| Browser Version | Google Chrome v100.0.4896.127 64-bit

# Actual Behavior

When, for any reason, the database-side deletion of media entities fails, the corresponding files are `unlink`ed nevertheless leading to erroneous behaviour (such as files being "not readable").

# Expected Behavior

The corresponding file should only be `unlink`ed if database-side deletion was successful (e.g. after `EntityManager::flush`).

# Steps to Reproduce

In my case it was a strict foreign key reference to a Media object:
1. Design a custom entity. Include a `ManyToOne` property referencing a `Media` entity, but omit cascade operations.
2. Create CRUD REST.
3. Create an entity and reference a `Media` object to it.
4. Try to delete the `Media` entity.

# Possible Solutions

Looking at `Sulu\Bundle\MediaBundle\Media\Manager\MediaManager:763`, file removal is apparently handled by the `FormatManager`:
```php
/** @var File $file */
foreach ($mediaEntity->getFiles() as $file) {
    /** @var FileVersion $fileVersion */
    foreach ($file->getFileVersions() as $fileVersion) {
        $this->formatManager->purge(
            $mediaEntity->getId(),
            $fileVersion->getName(),
            $fileVersion->getMimeType()
        );
        $this->storage->remove($fileVersion->getStorageOptions());
        foreach ($fileVersion->getMeta() as $fileVersionMeta) {
            // this will trigger massive-search deindex
            $this->em->remove($fileVersionMeta);
        }
        foreach ($fileVersion->getFormatOptions() as $formatOptions) {
            $this->em->detach($formatOptions);
        }
        $this->em->detach($fileVersion);
    }
    $this->em->detach($file);
}
```

Instead, purge data could be stored in a temp variable for deletion after `$this->em->flush()`:

```php
$toPurge = [];
$mediaId = $mediaEntity->getId();
/** @var File $file */
foreach ($mediaEntity->getFiles() as $file) {
    /** @var FileVersion $fileVersion */
    foreach ($file->getFileVersions() as $fileVersion) {
        $toPurge[] = [$fileVersion->getName(), $fileVersion->getMimeType()];
        $this->storage->remove($fileVersion->getStorageOptions());
        foreach ($fileVersion->getMeta() as $fileVersionMeta) {
            // this will trigger massive-search deindex
            $this->em->remove($fileVersionMeta);
        }
        foreach ($fileVersion->getFormatOptions() as $formatOptions) {
            $this->em->detach($formatOptions);
        }
        $this->em->detach($fileVersion);
    }
    $this->em->detach($file);
}

// ...

$this->em->flush();

foreach ($toPurge as $args) {
    $this->formatManager->purge($mediaId, $args[0], $args[1]);
}
```

**Comment by alexander-schranz** [comment]
The problem is that we don't want to remove the `Media` entity before we are not sure that all files are removed, so the behaviour you did describe is expected.

If we remove the files after we remove the entity, and removing the files because of some strange file system permission problem fails would be very bad.
The content manager could not longer remove that files later. So I think the current implemented way with deleting first the files and then delete the entity from the database is the best way. Sure the media will end in a state that it has no longer has a valid file for it. But that is better then just remove the entity from the database as the content manager would just ignore then the error as they not longer see the media in the list.

Specially in cases where somebody need to remove an image they not longer have a license to use it for this could lead to problems, when we make not sure that all files are removed before we removing it from our database.

At the end the error in your case is expected, and requires change on your side to change the column definition of the `ManyToOne` relation by adding `on-delete="SET NULL"` like here:

https://github.com/sulu/sulu/blob/a05e58a80cf83a40ed89981028cc03f3447caa86/src/Sulu/Bundle/ContactBundle/Resources/config/doctrine/Contact.orm.xml#L48

This will make sure that it does not throw a 500 error in your case when try to remove a media which is linked in a custom entity.

**Comment by R3dLantern** [comment]
I see your point. However, the system in question contains custom data where referenced media cannot be retroactively removed. I solved this by relaying `sulu_media.media_delete` requests to a custom controller that checks for custom data referring to the target media entity and throws HTTP 400 if necessary (that's where I also found the missing callback handler I fixed in #6591). I keep that solution for now. Thank you for the explanation 😃
